src/main/java/network/brightspots/rcv/CsvCvrReader.java

```diff
@@ -0,0 +1,111 @@
+import java.util.List;
+import java.util.Map;
+import javafx.util.Pair;
+import network.brightspots.rcv.RawContestConfig.CvrSource;
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
+
+class CsvCvrReader {
+  private final String cvrPath;
+  private final String contestId;
+  private final String undeclaredWriteInLabel;
+  private final ContestConfig contestConfig;
+  // 0-based column index of first ranking
+  private final int firstVoteColumnIndex;
+  // 0-based row index of first CVR
+  private final int firstVoteRowIndex;
+  // map for tracking unrecognized candidates during parsing
+  private final Map<String, Integer> unrecognizedCandidateCounts = new HashMap<>();
+
+  CsvCvrReader(
+      String cvrPath,
+      ContestConfig contestConfig,
+      CvrSource source) {
+    this.cvrPath = cvrPath;
+    this.contestConfig = contestConfig;
+    this.contestId = source.getContestId();
+    this.undeclaredWriteInLabel = source.getUndeclaredWriteInLabel();
+    this.firstVoteColumnIndex = Integer.parseInt(source.getFirstVoteColumnIndex()) - 1;
+    this.firstVoteRowIndex = Integer.parseInt(source.getFirstVoteRowIndex()) - 1;
+  }
+
+  // parse CVR CSV file into CastVoteRecord objects and add them to the input List<CastVoteRecord>
+  void readCastVoteRecords(List<CastVoteRecord> castVoteRecords)
+          throws CastVoteRecord.CvrParseException, IOException {
+    Logger.info("Reading CSV cast vote record file: %s...", cvrPath);
+
+    try (FileInputStream inputStream = new FileInputStream(Path.of(cvrPath).toFile())) {
+      CSVParser parser = CSVParser.parse(inputStream,
+              Charset.defaultCharset(), CSVFormat.newFormat(',').withHeader());
+      List<String> candidateIds = parser.getHeaderNames();
+      int undeclaredWriteInColumn = candidateIds.indexOf(undeclaredWriteInLabel);
+
+      parser.stream().skip(firstVoteRowIndex);
+
+      for (CSVRecord csvRecord : parser) {
+        ArrayList<Pair<Integer, String>> rankings = new ArrayList<>();
+        for (int col = firstVoteColumnIndex; col < csvRecord.size(); col++) {
+          String rankAsString = csvRecord.get(col);
+          if (rankAsString.isBlank()) {
+            continue;
+          }
+          int rankAsInt;
+          try {
+            rankAsInt = Integer.parseInt(rankAsString);
+          } catch (NumberFormatException e) {
+            Logger.severe("Row %s expected number at column %d, but got \"%s\" instead.",
+                    csvRecord.get(0), col, rankAsString);
+            throw new CastVoteRecord.CvrParseException();
+          }
+
+          String candidateId = candidateIds.get(col);
+          if (col == undeclaredWriteInColumn) {
+            candidateId = Tabulator.UNDECLARED_WRITE_IN_OUTPUT_LABEL;
+          } else if (!contestConfig.getCandidateCodeList().contains(candidateId)) {
+            unrecognizedCandidateCounts.merge(candidateId, 1, Integer::sum);
```

Comment:
I think you need to check this at the end and throw an error if it's > 0, as is done in `ClearBallotCvrReader` and the others as well.